### PR TITLE
hpctoolkit: overhaul to use spack dependencies

### PR DESCRIPTION
**Pull request by mwkrentel**
Overhaul hpctoolkit to AutotoolsPackage and to build its dependencies
directly from spack instead of using hpctoolkit-externals.

Remove hpctoolkit-externals since it's no longer needed.

--------------------------------------------------

This version should build and run everywhere the previous version
built.  It should run better due to using recent dyninst 10.0.

Probably neither will build on Cray due to spack's confusion of
front-end and back-end compilers as in issue #9613.  I don't even know
if there is a recipe that would build toolkit on Cray without some
tweaks inside spack.

Anyway, I'd prefer to update this now, get it into an all-spack build,
and then work on what's needed for Cray and Blue Gene.

ping @pramodskumbhar who wrote the previous recipe.  Maybe you'd like
to try it out, let me know if it works for your use case.

ping @scheibelp for review

### CODE REVIEW

File: var/spack/repos/builtin/packages/hpctoolkit/package.py [hunk 1]
```diff
@@ -17,31 +17,79 @@ class Hpctoolkit(Package):
...
+    variant('papi', default=False,
+            description='Use PAPI instead of perfmon for access to '
+            'the hardware performance counters.')
 
+    # We always support profiling MPI applications.  +mpi builds
+    # hpcprof-mpi, the MPI version of hpcprof.
+    variant('mpi', default=False,
+            description='Build hpcprof-mpi, the MPI version of hpcprof.')
+
+    variant('all-static', default=False,
```
[the remark this answers is not in the document]
Comment (mwkrentel) [reply var/spack/repos/builtin/packages/hpctoolkit/package.py]: No, in this case, all-static has nothing to do with shared libraries
or -fPIC.  +all-static adds the configure option --enable-all-static
which is the same as adding HPCPROFMPI_LT_LDFLAGS="-all-static" which
is needed for Cray or Blue Gene.  It's only for making hpcprof-mpi a
fully static binary.

File: var/spack/repos/builtin/packages/hpctoolkit/package.py [hunk 2]
```diff
@@ -17,31 +17,79 @@ class Hpctoolkit(Package):
...
+
+    boost_libs = '+atomic +graph +regex +serialization'  \
+        '+shared +multithreaded'
+
+    depends_on('binutils+libiberty~nls')
+    depends_on('boost' + boost_libs)
+    depends_on('bzip2')
+    depends_on('dyninst')
+    depends_on('elfutils~nls')
+    depends_on('intel-tbb')
```
[the remark this answers is not in the document]
Comment (mwkrentel) [reply var/spack/repos/builtin/packages/hpctoolkit/package.py]: No, tbb builds on all platforms that we support and is needed by
dyninst 10.x on all platforms.

The only issue for x86 is turning off transactional memory for AMD or
very old Intel machines that don't support it.

**Comment by mwkrentel** [comment]
@pramodskumbhar I've spent the last 3 months editing spack recipes and
the toolkit side so that this recipe, as simple and generic as it is
will work.

One of the main things was to eliminate special, toolkit-specific
patches that we relied on (especially for binutils).  As of my merge
on Nov 5 (d0c43e390), hpctoolkit master will accept much more generic
versions of our prereqs.

The plan is to replace externals with spack builds for the prereqs.
This won't be a sharp line, there will be a transition where both will
work for a time.

My plan is to support two workflows.  For developers, you can build
all the prereqs once with spack, let them sit somewhere, and then
build toolkit from that, the same way you use --with-externals now.
The machinery for this is already in place, but I'm not quite ready to
announce it.

For non-developer users, the plan is that 'spack install hpctoolkit'
should "just work," possibly with some variants or packages.yaml file
depending on your system.

As for versions of the prereqs, my plan is to keep hpctoolkit
package.py as generic as possible and to specify the versions and
variants in packages.yaml.  We'll publish the .yaml files in the
toolkit repo (but I'm not quite ready yet).

I was mainly interested in whether this recipe covers your use case,
in whatever spack environments or yaml files that you use.

**Comment by pramodk** [comment]
@mwkrentel  : fine then! just wanted to check if mainline development and spack recipes are/will be in sync.

**Pull request merged**